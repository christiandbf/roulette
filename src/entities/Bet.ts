import { strict as assert } from 'assert';
import Entity from './Entity';

interface Props {
  id?: string;
  rouletteId: string;
  userId: string;
  amount: number;
}

class Bet extends Entity<Props> {
  private readonly MAX_AMOUNT: number = 10000;

  private readonly rouletteId: string;
  private readonly userId: string;
  private readonly amount: number;

  constructor(props: Props) {
    super(props, props.id);
    assert.ok(props.amount < this.MAX_AMOUNT, 'Maximum bet amount exceeded');
    assert.ok(
      this.UUID_REGEXP.test(props.rouletteId),
      'User ID is not a valid UUID'
    );
    assert.ok(
      this.UUID_REGEXP.test(props.userId),
      'User ID is not a valid UUID'
    );
    this.rouletteId = props.rouletteId;
    this.userId = props.userId;
    this.amount = props.amount;
  }

  public getrouletteId(): string {
    return this.rouletteId;
  }

  public getUserId(): string {
    return this.userId;
  }

  public getAmount(): number {
    return this.amount;
  }
}

export default Bet;
