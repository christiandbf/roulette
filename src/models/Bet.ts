export interface BetRequestModel {
  id?: string;
  selection: string;
  rouletteId: string;
  userId: string;
  amount: number;
}

export interface BetResponseModel {
  id: string;
  selection: string;
  rouletteId: string;
  userId: string;
  amount: number;
}
