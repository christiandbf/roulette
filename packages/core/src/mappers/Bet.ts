import Mapper from './Mapper';
import Bet from '../domain/BetEntity';
import Option from '../domain/OptionValueObject';
import { BetRequestModel, BetResponseModel } from '../models/Bet';

const BetMapper: Mapper<Bet, BetRequestModel, BetResponseModel> = {
	toEntity: (request: BetRequestModel): Bet => {
		const option: Option = new Option({ selection: request.selection });
		return new Bet({
			option,
			amount: request.amount,
			rouletteId: request.rouletteId,
			userId: request.userId,
			id: request.id || undefined
		});
	},
	toModel: (bet: Bet): BetResponseModel => ({
		id: bet.getId(),
		rouletteId: bet.getRouletteId(),
		amount: bet.getAmount(),
		selection: bet.getOption().value,
		userId: bet.getUserId()
	})
};

export default BetMapper;
