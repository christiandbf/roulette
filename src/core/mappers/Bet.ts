import Mapper from './Mapper';
import Bet from '../domain/BetEntity';
import Option from '../domain/OptionValueObject';
import { BetRequestModel, BetResponseModel } from '../models/Bet';

const BetMapper: Mapper<Bet, BetRequestModel, BetResponseModel> = {
  toEntity: (request: BetRequestModel): Bet => {
    const option: Option = new Option({ selection: request.selection });
    const bet: Bet = new Bet({
      option,
      amount: request.amount,
      rouletteId: request.rouletteId,
      userId: request.userId
    });
    return bet;
  },
  toModel: (bet: Bet): BetResponseModel => ({
    id: bet.getId(),
    rouletteId: bet.getrouletteId(),
    amount: bet.getAmount(),
    selection: bet.getOption().value,
    userId: bet.getUserId()
  })
};

export default BetMapper;
