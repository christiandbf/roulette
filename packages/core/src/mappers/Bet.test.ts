import BetMapper from './Bet';
import Bet from '../domain/BetEntity';
import Option from '../domain/OptionValueObject';

describe('Bet mapper', () => {
	it('toEntity', () => {
		const entity = BetMapper.toEntity({
			amount: 1000,
			rouletteId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			selection: '1',
			userId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			id: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229'
		});
		expect(entity).toMatchSnapshot();
	});

	it('toEntity creating new ID', () => {
		const entity = BetMapper.toEntity({
			amount: 1000,
			rouletteId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			selection: '1',
			userId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			id: ''
		});
		expect(entity.getId()).toBeTruthy();
	});

	it('toModel', () => {
		const option = new Option({
			selection: '1'
		});
		const bet = new Bet({
			amount: 100,
			option: option,
			rouletteId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			userId: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229',
			id: 'ff6960b0-b015-11eb-8b4e-bd5c14d98229'
		});
		const model = BetMapper.toModel(bet);
		expect(model).toMatchSnapshot();
	});
});
