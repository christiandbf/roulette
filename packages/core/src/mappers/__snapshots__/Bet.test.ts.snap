// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bet mapper toEntity 1`] = `
Bet {
  "MAX_AMOUNT": 10000,
  "UUID_REGEXP": /\\^\\[0-9a-f\\]\\{8\\}-\\[0-9a-f\\]\\{4\\}-\\[0-5\\]\\[0-9a-f\\]\\{3\\}-\\[089ab\\]\\[0-9a-f\\]\\{3\\}-\\[0-9a-f\\]\\{12\\}\\$/i,
  "amount": 1000,
  "id": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
  "option": OptionValueObject {
    "SELECTION_REGEX": /\\^\\(\\[1-9\\]\\|\\[12\\]\\\\d\\|3\\[0-6\\]\\)\\$\\|\\^\\(BLACK\\|RED\\)\\$/i,
    "props": Object {
      "selection": "1",
    },
    "value": "1",
  },
  "props": Object {
    "amount": 1000,
    "id": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
    "option": OptionValueObject {
      "SELECTION_REGEX": /\\^\\(\\[1-9\\]\\|\\[12\\]\\\\d\\|3\\[0-6\\]\\)\\$\\|\\^\\(BLACK\\|RED\\)\\$/i,
      "props": Object {
        "selection": "1",
      },
      "value": "1",
    },
    "rouletteId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
    "userId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
  },
  "rouletteId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
  "userId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
}
`;

exports[`Bet mapper toModel 1`] = `
Object {
  "amount": 100,
  "id": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
  "rouletteId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
  "selection": "1",
  "userId": "ff6960b0-b015-11eb-8b4e-bd5c14d98229",
}
`;
